drivers/ethernet/eth_numaker.c

```diff
@@ -0,0 +1,815 @@
+/*
+ * Copyright (c) 2023 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_numaker_ethernet
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/reset.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/clock_control_numaker.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/net/ethernet.h>
+#include "eth_numaker_priv.h"
+#include "ethernet/eth_stats.h"
+#include <soc.h>
+#include <NuMicro.h>
+#include <synopGMAC_network_interface.h>
+
+#ifdef CONFIG_SOC_M467
+#include <m460_eth.h>
+#endif
+
+LOG_MODULE_REGISTER(eth_numaker, CONFIG_ETHERNET_LOG_LEVEL);
+
+/* Device EMAC Interface port */
+#define NUMAKER_GMAC_INTF  0
+/* 2KB Data Flash at 0xFF800 */
+#define NUMAKER_DATA_FLASH (0xFF800U)
+#define NUMAKER_MASK_32    (0xFFFFFFFFU)
+extern synopGMACdevice GMACdev[GMAC_CNT];
+extern struct sk_buff tx_buf[GMAC_CNT][TRANSMIT_DESC_SIZE];
+extern struct sk_buff rx_buf[GMAC_CNT][RECEIVE_DESC_SIZE];
+
+static uint32_t eth_phy_addr;
+
+/* Device config */
+struct eth_numaker_config {
+	/* GMAC Base address */
+	uint32_t gmac_base;
+	const struct reset_dt_spec reset;
+	uint32_t phy_addr;
+	/* clock configuration */
+	uint32_t clk_modidx;
+	uint32_t clk_src;
+	uint32_t clk_div;
+	const struct device *clk_dev;
+	const struct pinctrl_dev_config *pincfg;
+};
+
+/* Driver context/data */
+struct eth_numaker_data {
+	synopGMACdevice *gmacdev;
+	struct net_if *iface;
+	uint8_t mac_addr[6];
+	struct k_mutex tx_frame_buf_mutex;
+	struct k_spinlock rx_frame_buf_lock;
+};
+
+/* Delay execution for given amount of ticks for SDK-HAL */
+void plat_delay(uint32_t delay)
+{
+	int64_t start = k_uptime_ticks();
+
+	while (1) {
+		if ((k_uptime_ticks() - start) >= delay) {
+			break;
+		}
+	}
```

Comment:
There is a `k_busy_wait()` that could be used instead.

Reply:
How could I convert ticks to usec ?

Reply:
k_ticks_to_us_floor32 could convert to usec.
I applied k_busy_wait().